### PR TITLE
drain manager: Make probabilistic drain configurable

### PR DESCRIPTION
**Pull request by auni53**
* Add DrainStrategy enum to Options with `Graceful` and `Immediate`
* Disable probabilistic drain in DrainManager if DrainStrategy == Immediate
* Add integration tests

Risk Level: Low.
Testing: Integration tests, verify that the race condition from #11240 does not occur if the probabilistic drain is disabled.

### CODE REVIEW

**Comment by auni53** [comment]
@mattklein123 wanted to get your thoughts on this before I add the docs and release notes. I added a bool that guards the probabilistic drain behaviour, but I've got two thoughts:

1. Instead of `bool drainIncrementally()` I was thinking about `enum drainStrategy()`, where we can define different drain strategies in the server options header. There'd just be two right now, but this would make it easier to put more of the overall server shutdown sequence into the drainmanager in the future.

2. It's a little tiresome having to change so many codepaths every time we want to add a new option to the integration server config. Is it possible to modify the server optionsimpl directly? If not, I'll probably add a simple `struct DrainManagerOptions` to pass in the three drain manager configs.

**Comment by mattklein123** [comment]
> I was thinking about enum drainStrategy(), where we can define different drain strategies in the server options header. There'd just be two right now, but this would make it easier to put more of the overall server shutdown sequence into the drainmanager in the future.

+1 agreed this seems more future proof.

> It's a little tiresome having to change so many codepaths every time we want to add a new option to the integration server config. Is it possible to modify the server optionsimpl directly? If not, I'll probably add a simple struct DrainManagerOptions to pass in the three drain manager configs.

For integration tests it seems reasonable to allow direct access, but I will defer to @alyssawilk on that (it might already be possible).

**Comment by alyssawilk** [comment]
I don't mind allowing direct access, but please be mindful of thread safety issues - I think accessing it after the server starts running would be problematic.

**Comment by auni53** [comment]
> I don't mind allowing direct access, but please be mindful of thread safety issues - I think accessing it after the server starts running would be problematic.

I decided this was too complicated for this PR and just plumbed the other variable. I'll maybe try to get this to work on the side.

**Comment by alyssawilk** [comment]
I'm going to assume this is still not ready for review as CI is red.  Ping when it's ready for a pass!

**Comment by auni53** [comment]
OK @alyssawilk this is green.

File: source/server/options_impl.h [hunk 1]
```diff
@@ -65,12 +65,13 @@ class OptionsImpl : public Server::Options, protected Logger::Loggable<Logger::I
...
   }
   void setDrainTime(std::chrono::seconds drain_time) { drain_time_ = drain_time; }
-  void setLogLevel(spdlog::level::level_enum log_level) { log_level_ = log_level; }
-  void setLogFormat(const std::string& log_format) { log_format_ = log_format; }
-  void setLogPath(const std::string& log_path) { log_path_ = log_path; }
   void setParentShutdownTime(std::chrono::seconds parent_shutdown_time) {
     parent_shutdown_time_ = parent_shutdown_time;
   }
+  void setDrainStrategy(Server::DrainStrategy drain_strategy) { drain_strategy_ = drain_strategy; }
+  void setLogLevel(spdlog::level::level_enum log_level) { log_level_ = log_level; }
```
Comment (alyssawilk) [inline source/server/options_impl.h:72]: Why move these?

File: test/server/drain_manager_impl_test.cc [hunk 2]
```diff
@@ -98,24 +128,27 @@ TEST_F(DrainManagerImplTest, DrainDeadlineProbability) {
...
 }
 
-TEST_F(DrainManagerImplTest, ModifyOnly) {
-  InSequence s;
-  DrainManagerImpl drain_manager(server_, envoy::config::listener::v3::Listener::MODIFY_ONLY);
-
-  EXPECT_CALL(server_, healthCheckFailed()).Times(0); // Listener check will short-circuit
-  EXPECT_FALSE(drain_manager.drainClose());
-}
+INSTANTIATE_TEST_SUITE_P(DrainGradually, DrainManagerImplTest, testing::Bool());
```
Comment (alyssawilk) [inline test/server/drain_manager_impl_test.cc:151]: Is DrainGradually the correct name for this?  Perhaps DrainStrategies?

File: test/integration/integration.h [hunk 3]
```diff
@@ -436,6 +436,8 @@ class BaseIntegrationTest : protected Logger::Loggable<Logger::Id::testing> {
   // The duration of the drain manager graceful drain period.
   std::chrono::seconds drain_time_{1};
 
+  Server::DrainStrategy drain_strategy_{Server::DrainStrategy::Gradual};
```
Comment (alyssawilk) [inline test/integration/integration.h:439]: Comments encouraged.

File: source/server/options_impl.cc [hunk 4]
```diff
@@ -387,9 +387,9 @@ OptionsImpl::OptionsImpl(const std::string& service_cluster, const std::string&
       log_format_(Logger::Logger::DEFAULT_LOG_FORMAT), log_format_escaped_(false),
       restart_epoch_(0u), service_cluster_(service_cluster), service_node_(service_node),
       service_zone_(service_zone), file_flush_interval_msec_(10000), drain_time_(600),
-      parent_shutdown_time_(900), mode_(Server::Mode::Serve), hot_restart_disabled_(false),
-      signal_handling_enabled_(true), mutex_tracing_enabled_(false), cpuset_threads_(false),
-      fake_symbol_table_enabled_(false) {}
+      parent_shutdown_time_(900), drain_strategy_(Server::DrainStrategy::Gradual),
+      mode_(Server::Mode::Serve), hot_restart_disabled_(false), signal_handling_enabled_(true),
```
Comment (alyssawilk) [inline source/server/options_impl.cc:391]: How are non-google users to configure this option?

File: source/server/options_impl.h (hunk 1 above)
Comment (auni53) [reply source/server/options_impl.h:72]: Haha, the diff makes it look weird, but I was trying to move `setParentShutdownTime(std::chrono::seconds)` in order to group all of the drain-related setters together.

File: test/server/drain_manager_impl_test.cc (hunk 2 above)
Comment (auni53) [reply test/server/drain_manager_impl_test.cc:151]: Done.

File: source/server/options_impl.cc (hunk 4 above)
Comment (auni53) [reply source/server/options_impl.cc:391]: Assuming you're saying this config path is only used by Google?

I was sticking to YAGNI, don't add further complications to the config plane unless someone actually wants it. I'd be happy to add CLI options for this similar to the --drain-time-s, though I'd prefer to do that in a separate PR.

Comment (alyssawilk) [reply source/server/options_impl.cc:391]: It's pretty envoy-standard to land the configuration, release notes, and code changes in a single PR - I don't think it'll make this one too large to review.

Comment (auni53) [reply source/server/options_impl.cc:409]: Done I think, I looked at similar PRs to try to make sure I updated the right spot everywhere, but I'm not sure if I missed anything or if the language is correct.

File: api/envoy/admin/v3/server_info.proto [hunk 5]
```diff
@@ -75,6 +75,14 @@ message CommandLineOptions {
     InitOnly = 2;
   }
 
+  enum DrainStrategy {
```
Comment (htuch) [inline api/envoy/admin/v3/server_info.proto:78]: Should this be a bootstrap or Listener option or CLI?

Comment (htuch) [reply api/envoy/admin/v3/server_info.proto:78]: Sure, the consistency argument is strong.

**Pull request merged**